k2/csrc/intersect.cu

```diff
@@ -90,950 +69,777 @@ using namespace intersect_internal;  // NOLINT
+      // i is actually an idx012
+
+      // the following works because each FSA has (its num-states * T_+1) states
+      // allocated to it.  However (i / (T_+1)) does not directly map to a state
+      // index.
+      int32_t fsa_idx0 = a_fsas_row_splits1_data[(i / (T+1))];
+      FsaInfo fsa_info = fsa_info_data[fsa_idx0];
+      float cutoff = score_cutoffs_data[fsa_idx0];
+
+      int32_t idx_within_fsa = i - (T+1) * fsa_info.state_offset,
+                    t = idx_within_fsa / fsa_info.num_states,
+           state_idx1 = idx_within_fsa % fsa_info.num_states,
+           state_idx01 = fsa_info.state_offset + state_idx1;
+
+      char keep = (char)0;
+      if (t <= fsa_info.T) {
+        // This time is within the bounds for this FSA
+        const float *forward_state_scores_t = forward_state_scores_data[t],
+                   *backward_state_scores_t = backward_state_scores_data[t];
+        if (forward_state_scores_t[state_idx01] +
+            backward_state_scores_t[state_idx01] > cutoff)
+          keep = (char)1;
       }
-      char keep_this_state =
-          (backward_loglike + forward_loglike >= -output_beam);
-
-      // we can use the arcs row-splits because the structure of
-      // FrameInfo::states is the same as the top level structure of
-      // FrameInfo::arcs.
-      int32_t states_idx0x = arcs_row_splits1[states_idx0],
-              states_idx1 = state_idx01 - states_idx0x;
-
-      int32_t oshape_idx0x = oshape_row_splits1[states_idx0],
-              oshape_idx01 = oshape_idx0x + t,
-              oshape_idx01x = oshape_row_splits2[oshape_idx01],
-              oshape_idx012 = oshape_idx01x + states_idx1;
-      // note: axis 1 of 'states' corresponds to axis 2 of 'oshape'; it's the
-      // state index.  Also,
-
-      keep_states_data[oshape_idx012] = keep_this_state;
-      if (!keep_this_state) {
-        // The reason we set the backward_loglike to -infinity here if it's
-        // outside the beam, is to prevent disconnected states from appearing
-        // after pruning due to numerical roundoff effects near the boundary
-        // at `-beam`.  It would otherwise be correct and harmless to omit
-        // this if-block.
-        backward_loglike = minus_inf;
+      keep_state_data[i] = keep;
+    };
+    Eval(c_, product, lambda_set_keep);
+
+    Array1<int32_t> &new2old = renumber_states.New2Old();
+    const int32_t *new2old_data = new2old.Data();
+    int32_t ans_tot_num_states = new2old.Dim();
+
+    // t_per_fsa will be set below to the number of time-steps that each FSA has
+    // states active on; if each FSA i has scores for 0 <= t < T_i, then
+    // t_per_fsa[i] will be T_i + 1, because there is also a copy of the state
+    // at time T_i.
+    Array1<int32_t> t_per_fsa(c_, num_fsas_ + 1);
+    int32_t *t_per_fsa_data = t_per_fsa.Data();
+    auto lambda_set_t_per_fsa_etc = [=] __host__ __device__ (int32_t i) -> void {
+      t_per_fsa_data[i] = fsa_info_data[i].T + 1;
```

Comment:
This is a very short lambda, maybe we should change it to `if CPU else lambda` approach.

Reply:
So we're still getting that problem with short lambdas?  We didn't somehow resolve it another way?

Reply:
I supposed we just try our best to avoid short lambdas nowdays.